### PR TITLE
[GRPO] Add metrics for low and high clipped token probabilities

### PR DESCRIPTION
**Pull request by lewtun**
# What does this PR do?

This PR adds metrics to track low and high clipped token probabilities independently. It is useful for finding a suitable empirical value of `ε_high` which cannot be done from `clip_ratio` alone.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1086,12 +1086,30 @@ def _compute_loss(self, model, inputs):
...
+        is_region_clipped = is_low_clipped | is_high_clipped
+
+        low_clip = (is_low_clipped * completion_mask).sum() / completion_mask.sum()
+        high_clip = (is_high_clipped * completion_mask).sum() / completion_mask.sum()
+        clip_ratio = (is_region_clipped * completion_mask).sum() / completion_mask.sum()
+
+        self._metrics[mode]["clip_ratio/low_mean"].append(
+            self.accelerator.gather_for_metrics(low_clip).nanmean().item()
+        )
+        self._metrics[mode]["clip_ratio/low_min"].append(
```
Comment (lewtun) [inline trl/trainer/grpo_trainer.py:1101]: I don't include `low_max` since we only care about the extrema of the clipping

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1086,12 +1086,30 @@ def _compute_loss(self, model, inputs):
...
+            torch.min(self.accelerator.gather_for_metrics(low_clip)).item()
+        )
+        self._metrics[mode]["clip_ratio/high_mean"].append(
+            self.accelerator.gather_for_metrics(high_clip).nanmean().item()
+        )
+        self._metrics[mode]["clip_ratio/high_max"].append(
```
Comment (lewtun) [inline trl/trainer/grpo_trainer.py:1107]: Ditto for high

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -1086,12 +1086,30 @@ def _compute_loss(self, model, inputs):
             mean_kl = (per_token_kl * completion_mask).sum() / completion_mask.sum()
             self._metrics[mode]["kl"].append(self.accelerator.gather_for_metrics(mean_kl).nanmean().item())
 
-        # Compute the clip ratio
-        is_clipped = ((coef_1 < 1 - self.epsilon_low) & (advantages.unsqueeze(1) < 0)) | (
-            (coef_1 > 1 + self.epsilon_high) & (advantages.unsqueeze(1) > 0)
+        # Compute the clipped probability ratios
```
Comment (lewtun) [inline trl/trainer/grpo_trainer.py:1089]: @kashif do I also need to update the Liger loss to account for these new metrics?

**Review by edbeeching — APPROVED** [review]

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -1086,12 +1086,30 @@ def _compute_loss(self, model, inputs):
...
+            self.accelerator.gather_for_metrics(low_clip).nanmean().item()
+        )
+        self._metrics[mode]["clip_ratio/low_min"].append(
+            torch.min(self.accelerator.gather_for_metrics(low_clip)).item()
+        )
+        self._metrics[mode]["clip_ratio/high_mean"].append(
+            self.accelerator.gather_for_metrics(high_clip).nanmean().item()
+        )
+        self._metrics[mode]["clip_ratio/high_max"].append(
+            torch.max(self.accelerator.gather_for_metrics(high_clip)).item()
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1108]: Some values in the gathered `high_clip` can be `nan`, (eg, because of the overlong filtering). We need a `nanmax` here that would ignore these values. Currently it would work, but as soon as you have one nan, the logged value is nan.
I'll push a fix so that you can merge tomorrow

Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: Red: before e6d013f
Green: after e6d013f

Screenshot 2025-04-15 at 14 39 51

Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]:
```python
from datasets import load_dataset
from trl import GRPOTrainer, GRPOConfig
import random

dataset = load_dataset("trl-lib/tldr", split="train[:200]")


def reward_random(completions, **kwargs):
    return [random.random() for _ in range(len(completions))]

trainer = GRPOTrainer(
    model="Qwen/Qwen2.5-1.5B",
    reward_funcs=reward_random,
    train_dataset=dataset,
    args=GRPOConfig(
        num_iterations=4,
        per_device_train_batch_size=2,
        mask_truncated_completions=True,
        logging_steps=10,
    ),
)
trainer.train()
```

4 GPUs

```sh
accelerate launch 3289.py
```

**Review by qgallouedec — APPROVED** [review]
LGTM!

File: trl/trainer/grpo_trainer.py (hunk 4 above)
Comment (lewtun) [reply trl/trainer/grpo_trainer.py]: Very nice fix, thanks!
